File: go/pointers/binary_search_tree.go
package main

import "fmt"

type TreeNode struct {
    value int
    left *TreeNode
    right *TreeNode
}

func (node *TreeNode) insert(value int) {
    if value <= node.value {
        if node.left == nil {
            node.left = &TreeNode{value: value}
        } else {
            node.left.insert(value)
        }
    }
}

func (node *TreeNode) inOrder() {
    if node == nil {
        return
    }
    node.left.inOrder()
    fmt.Print(node.value, " ") 
    node.right.inOrder()
}

func main() {
	root := &TreeNode{value: 10}
	values := []int{5, 15, 3, 7, 12, 18}

	for _, v := range values {
		root.insert(v)
	}

	fmt.Print("In-order traversal: ")
	root.inOrder()
	fmt.Println()
}

File: rust/binary-tree/binary_search_tree.rs
use std::comp::Ordering;

struct Node<T> {
    value: T,
    left: Option<Box<Node<T>>>,
    right: Oprion<Box<Node<T>>>,
}

impl<T:Ord + std::fmt::Display> Node<T> {
    fn new(value: T) -> Self {
        Node {
            value,
            left: None,
            right: None,
        }
    }

    fn insert (&mut self, new_value: T) {
        match new_value.cmp(&self.value) {
            Ordering::Less => {
                if let Some(ref mut left) = self.left {
                    left.insert(new_value);
                } else {
                    self.left = Some(Box::new(Node::new(new_value)));
                }
            }
            Ordering::Greater => {
                if let Some(ref mut right) = self.right {
                    right.insert(new_value);
                } else {
                    self.right = Some(Box::new(Node::new(new_value)));
                }
            }
            Ordering::Equal => {
                // Do nothing if the new value is equal (no duplicates allowed).
            }
        }
    }

    fn contains(&self, target_value: &T) -> bool {
        match target_value.cmp(&self.value) {
            Ordering::Less => {
                if let Some(ref left) = self.left {
                    left.contains(target_value)
                } else {
                    false
                }
            }
        }
    }
}
